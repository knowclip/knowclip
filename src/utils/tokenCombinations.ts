//    "START-TOKEN-BASED" improved?
// maybe sort by length makes sense.
// first individual words matches.
//         1   das
//         -   das
// then:   5   das licht spiegelt sich in
// no limit-   das licht spiegelt sich in
//         4   das licht spiegelt sich
//         -   das licht spiegelt sich
//         4   das licht spiegelt      in
//         -   das licht spiegelt      in
//         4   das licht          sich in
//         -   das licht          sich in
//         4   das       spiegelt sich in
//         -   das       spiegelt sich in
//         3   das licht spiegelt
//         -   das licht spiegelt
//         3   das licht          sich
//         -   das licht          sich
//         3   das licht               in
//         -   das licht               in
//         3   das       spiegelt sich
//         -   das       spiegelt sich
//         3   das       spiegelt      in
//         -   das       spiegelt      in
//         3   das                sich in
//         2   das licht
//         -   das licht
//         2   das       spiegelt
//         -   das       spiegelt
//         2   das                sich
//         -   das                sich
//         2   das                     in
//         -   das                     in

export function getTokenCombinations<T>(tokens: T[]) {
  return [
    ...getCombinations(tokens.slice(1))
      .sort((a, b) => b.length - a.length)
      .map((ts) => [tokens[0], ...ts]),
    [tokens[0]],
  ]
}

function getCombinations<T>(items: T[]) {
  const result: T[][] = []
  const f = function (prefix: T[], items: T[]) {
    for (let i = 0; i < items.length; i++) {
      result.push([...prefix, items[i]])
      f([...prefix, items[i]], items.slice(i + 1))
    }
  }
  f([], items)
  return result
}
