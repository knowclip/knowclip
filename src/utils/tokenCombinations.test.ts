import { getTokenCombinations } from './tokenCombinations'
import { describe, it, expect } from 'vitest'

describe('getTokenCombinations', () => {
  it('gives all combinations for five items in right order', () => {
    expect(
      getTokenCombinations(['das', 'licht', 'spiegelt', 'sich', 'in'])
    ).toMatchSnapshot()
  })
})
