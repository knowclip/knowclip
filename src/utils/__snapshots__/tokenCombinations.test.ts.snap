// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getTokenCombinations > gives all combinations for five items in right order 1`] = `
[
  [
    "das",
    "licht",
    "spiegelt",
    "sich",
    "in",
  ],
  [
    "das",
    "licht",
    "spiegelt",
    "sich",
  ],
  [
    "das",
    "licht",
    "spiegelt",
    "in",
  ],
  [
    "das",
    "licht",
    "sich",
    "in",
  ],
  [
    "das",
    "spiegelt",
    "sich",
    "in",
  ],
  [
    "das",
    "licht",
    "spiegelt",
  ],
  [
    "das",
    "licht",
    "sich",
  ],
  [
    "das",
    "licht",
    "in",
  ],
  [
    "das",
    "spiegelt",
    "sich",
  ],
  [
    "das",
    "spiegelt",
    "in",
  ],
  [
    "das",
    "sich",
    "in",
  ],
  [
    "das",
    "licht",
  ],
  [
    "das",
    "spiegelt",
  ],
  [
    "das",
    "sich",
  ],
  [
    "das",
    "in",
  ],
  [
    "das",
  ],
]
`;
